### PR TITLE
fix(queue): recover non-capture LlmRequests stuck in Processing (#1209)

### LINKED ISSUE #1209: No sweeper recovers non-capture LlmRequests stuck in Processing
## Problem

After PR #1200 (issue #1190) made non-capture LlmQueue claims atomic via `TryClaimProcessingAsync`, there is a recovery gap: if a worker crashes (or its process is killed) **after** successfully claiming a non-capture `LlmRequest` (flipping it to `Processing`) but **before** completing/failing it, the row is left in `Processing` **forever**. Nothing transitions it back to `Pending` or `Failed`, so the item is silently abandoned.

This was acknowledged as the residual half of review finding M1 on PR #1200 ("orphaned Processing item"), which was mitigated for the re-fetch-null branch but not for the broader worker-crash-after-claim case.

## Why nothing recovers it today

- `ProposalHousekeepingWorker` only operates on proposals -- it has **zero** references to `LlmRequest` / `RequestStatus.Processing`, so it does not sweep stuck queue items.
- Capture-triage items are read from `Processing` each poll and re-claimed, so they self-heal; **non-capture** items are read from `Pending` only, so once stuck in `Processing` they are never re-enqueued.
- The `LlmQueueToProposalWorker` itself has no stuck-item reclamation pass.

## Existing pattern to mirror

`OutboundWebhookDeliveryWorker.RecoverStuckProcessingDeliveriesAsync` (backend/src/Taskdeck.Api/Workers/OutboundWebhookDeliveryWorker.cs ~line 284) already implements exactly this shape: query rows that have sat in `Processing` past a threshold (`GetStuckProcessingAsync`) and `ReturnToPending(...)` them. A parallel sweeper for stuck `LlmRequest` rows (e.g. in `ProposalHousekeepingWorker` or `LlmQueueToProposalWorker`) should:

1. Query non-capture `LlmRequest`s in `Processing` whose `UpdatedAt` is older than a configurable stuck threshold.
2. Return them to `Pending` (respecting retry budget) or mark `Failed` when retries are exhausted.
3. Emit telemetry for recovered items.

## Acceptance criteria

- A background sweeper recovers non-capture `LlmRequest`s stuck in `Processing` beyond a configurable threshold.
- Retry budget is respected (return to `Pending` until max retries, then `Failed`).
- Recovery is covered by tests (worker crash after claim -> item eventually reclaimed).
- Threshold is configurable via `WorkerSettings` and documented in `docs/platform/CONFIGURATION_REFERENCE.md`.

## References
- PR #1200 (atomic non-capture claim)
- Issue #1190 (queue claim race)
- Pattern: `OutboundWebhookDeliveryWorker.RecoverStuckProcessingDeliveriesAsync`

### PR DESCRIPTION
**Pull request by Chris0Jeky**
## Summary

Fixes #1209. A non-capture `LlmRequest` claimed for processing (flipped to `Processing`) by a worker that then **crashed before completing or failing it** was left in `Processing` forever — capture-triage items self-heal because they are read from `Processing` and re-claimed every poll, but non-capture (automation) items are read **only from `Pending`**, so once stuck they were silently abandoned. This was the residual half of finding M1 on #1200.

## What changed

- **`LlmQueueToProposalWorker`** runs a recovery pass (`RecoverStuckProcessingItemsAsync`) at the top of each `ProcessBatchAsync` tick, mirroring the shipped `OutboundWebhookDeliveryWorker.RecoverStuckProcessingDeliveriesAsync` pattern the issue cites. A recovered item becomes eligible again within the same tick.
- **`ILlmQueueRepository.GetStuckProcessingNonCaptureAsync(staleBefore, limit)`** — new dual SQLite-raw / EF query returning non-capture rows in `Processing` whose `UpdatedAt <= staleBefore`, oldest-first, bounded. Capture rows are excluded in-query (they self-heal). The `UpdatedAt <= staleBefore` shape matches the shipped webhook stuck-query.
- **Retry budget respected:** each stuck item is `MarkAsFailed` (RetryCount++) then `ResetForRetry` back to `Pending` **while budget remains** (`RetryCount + 1 < MaxRetries`, the same check as `HandleFailureWithRetryAsync`), else left `Failed`. A request that repeatedly crashes the worker therefore exhausts its budget and fails permanently instead of looping forever.
- **Threshold** = `Math.Max(30, Workers:ProcessingLeaseSeconds)` (reused setting, like the webhook sweep; the 30s floor stops an aggressively-low lease from reclaiming an in-flight request). Documented in `docs/platform/CONFIGURATION_REFERENCE.md`.
- **Telemetry:** recovered/failed counts via `WorkerItemsProcessed` (`outcome=recovered_stuck_requeued` / `recovered_stuck_failed`) + a warning log.

## Acceptance criteria

- [x] Background sweeper recovers non-capture rows stuck in `Processing` beyond a configurable threshold.
- [x] Retry budget respected (Pending until max retries, then Failed).
- [x] Covered by tests (worker crash after claim -> eventually reclaimed).
- [x] Threshold configurable via `WorkerSettings` + documented in `CONFIGURATION_REFERENCE.md`.

## Testing

- **Worker tests** (`LlmQueueToProposalWorkerTests`): stuck item -> Pending + budget consumed; budget-exhausted -> Failed; fresh `Processing` item not swept; stuck **capture** item not swept; **end-to-end** tick recovers then reprocesses a stuck item to `Completed`.
- **Repository integration tests** (real SQLite): selects only non-capture `Processing` rows at/before the threshold; excludes rows newer than the threshold; rejects `limit < 1`.
- Local (Release): targeted Api.Tests (worker + repo + resilience) 74/74; Application.Tests LlmQueue-related 96/96; Architecture.Tests 20/21 (1 pre-existing skip); backend build clean. Full suite runs in CI (the full Api.Tests suite stack-overflows locally on Windows under parallelism — environmental, unrelated; the implicated `AuditRetention` class passes in isolation).

## Notes

- Adding an interface method updated the 3 hand-written `ILlmQueueRepository` fakes; Moq mocks auto-stub.
- The >lease legitimate-processing double-run window is inherent to lease-based recovery and pre-exists in the webhook worker; review-first proposals bound any worst case. Not newly introduced.

### CODE REVIEW

**Comment by Chris0Jeky** [comment]
## Two independent adversarial reviews — findings + fixes

Ran two independent adversarial reviews with non-overlapping lenses (correctness/concurrency and test-fidelity/regression). The correctness review found **no CRITICAL/HIGH** (verdict: safe); the test review found two **HIGH** coverage gaps. Per the repo review policy, **every finding at every severity is fixed below** (pushed in `efa0e2d0` + `7e1843f1`).

### Correctness / concurrency lens
| # | Sev | Finding | Resolution |
|---|-----|---------|------------|
| C1 | LOW–MED | Recovery sweep was the first awaited call in `ProcessBatchAsync` with no local try/catch — a recovery `SaveChanges` failure would skip the tick's normal queue draining. | **Fixed** `efa0e2d0`: wrapped the sweep in its own try/catch (re-throws `OperationCanceledException` on shutdown; logs + continues otherwise), so a recovery hiccup can't starve draining. |
| C2 | LOW | Comment misattributed the SQLite `UpdatedAt` TEXT-comparison correctness to ADR-0040 (which scopes only `DateTime` Kind normalization, not `DateTimeOffset`). | **Fixed** `efa0e2d0`: comment now explains the real reason — all writers + `staleBefore` use `DateTimeOffset.UtcNow` (uniform `+00:00`), so lexical order == chronological, matching the shipped webhook stuck-query. |
| C3 | LOW | Comment overstated "consistent with in-band failure" — recovery omits the `IsTransientFailure` gate. | **Fixed** `efa0e2d0`: comment now states a crash is always treated as retryable (no error code to classify) and documents the single-worker assumption + lease-based spacing for poison-pills. |
| C4 | LOW | Multi-instance double-recovery (no optimistic-concurrency guard). | Pre-existing/by-design for the local-first single-worker model; mirrors the webhook recovery sweep. Documented in code (`efa0e2d0`) and the PR body. Verified single-process can't sweep its own in-flight work (sweep runs before, and is awaited sequentially with, the drain). |
| C5 | LOW | Poison-pill requeue has no explicit backoff. | Bounded by the retry budget; the lease itself spaces successive recoveries ≥ lease apart. Documented; no blocking delay added (would harm the normal path). |
| — | — | Retry-budget math, domain transitions, change-tracking persistence, capture/non-capture predicate, per-tick cost. | **Verified correct, no change needed.** |

### Test-fidelity / regression lens
| # | Sev | Finding | Resolution |
|---|-----|---------|------------|
| T1 | HIGH | Oldest-first `ORDER BY` + `LIMIT` unproven on real data — the integration test used the order-insensitive `BeEquivalentTo` and `limit > rows`, so the in-memory re-sort guard and the bound had zero coverage. Risk: a wrong order could drop the **oldest** abandoned rows. | **Fixed** `7e1843f1`: new integration test inserts 3 rows with **distinct backdated `UpdatedAt`**, queries `limit: 2`, asserts ordered `.Equal(oldest, middle)` (newest excluded). |
| T2 | HIGH | AC4 (configurable threshold) + the 30s floor were vacuously covered — every test used the 120s default and `-10000s` ages, so the lease/floor branch was never exercised. | **Fixed** `7e1843f1`: a `[Theory]` straddles the effective lease at `(lease=10→floor 30, age 20→not swept / age 40→swept)` and `(lease=300, age 150→not swept / age 400→swept)`. |
| T3 | MED | Worker lease boundary only coarsely tested. | **Fixed** by the T2 theory (near-boundary cases). |
| T4 | MED | Recovery lacks the claim path's optimistic-concurrency guard; no double-sweep test. | **Fixed** `7e1843f1`: idempotency test proves a second sweep of an already-requeued item doesn't double-consume budget. Single-worker assumption documented (C4). |
| T5 | LOW | Telemetry emission untested. | **Acknowledged, intentionally skipped** — consistent with the entire codebase (no worker-telemetry has test coverage; a `MeterListener` harness would add fragile, precedent-less infra). Reviewer agreed this is acceptable. |
| T6 | LOW | No worker-level mixed capture+non-capture partition test. | **Fixed** `7e1843f1`: added a mixed worker test (only non-capture swept). |
| — | — | Fakes faithfully model real behavior; no regression to existing worker/resilience tests; no Moq-null NPE; all `ILlmQueueRepository` fakes updated. | **Verified, no change needed.** |

### Verification after fixes (local, Release)
- Backend build clean.
- Targeted Api.Tests (`LlmQueueToProposalWorkerTests` + `LlmQueueRepositoryIntegrationTests` + both `WorkerResilience`): **81/81** (was 74; +7 gap-closing tests).
- Full backend suite runs in CI (the full Api.Tests suite stack-overflows locally on Windows under parallelism — environmental; the implicated `AuditRetention` class passes in isolation, and CI's Backend Unit is green on both ubuntu and windows).

File: backend/src/Taskdeck.Api/Workers/LlmQueueToProposalWorker.cs [hunk 1]
```diff
@@ -64,6 +69,12 @@ protected override async Task ExecuteAsync(CancellationToken stoppingToken)
 
     private async Task ProcessBatchAsync(CancellationToken ct)
     {
+        // Reclaim non-capture requests abandoned in Processing by a crashed worker before draining the
+        // queue (#1209), so a recovered item becomes eligible again within this same tick. Capture-triage
+        // items self-heal via the Processing re-claim path below, so the sweep targets non-capture work
+        // only -- the one kind read solely from Pending and otherwise stuck forever.
+        await RecoverStuckProcessingItemsAsync(ct);
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Api/Workers/LlmQueueToProposalWorker.cs]: Fixed in `efa0e2d0`: the recovery sweep is now wrapped in its own try/catch (re-throws `OperationCanceledException` on shutdown; logs + continues otherwise), so a recovery failure can never starve the tick's normal queue draining. Matches this suggestion, plus clean-shutdown handling.

File: backend/src/Taskdeck.Infrastructure/Repositories/LlmQueueRepository.cs [hunk 2]
```diff
@@ -309,6 +309,47 @@ private async Task<int> CountByStatusAndCaptureKindAsync(
...
+            // FromSqlInterpolated + Include wraps this in a subquery that does not guarantee the raw
+            // ORDER BY survives (see GetByUserAsync); the inner LIMIT still selects the correct oldest-N
+            // rows, so re-sort oldest-first in memory to make the ordering a contract.
+            var rows = await _context.LlmRequests
+                .FromSqlInterpolated(
+                    $"SELECT * FROM LlmRequests WHERE Status = {(int)RequestStatus.Processing} AND RequestType NOT LIKE {CaptureRequestTypeLike} AND UpdatedAt <= {staleBefore} ORDER BY UpdatedAt ASC LIMIT {limit}")
+                .Include(lr => lr.User)
+                .Include(lr => lr.Board)
+                .ToListAsync(cancellationToken);
+            return rows.OrderBy(lr => lr.UpdatedAt).ToList();
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Infrastructure/Repositories/LlmQueueRepository.cs:339]: Fixed in `6cab6881`: dropped `.Include(User)`/`.Include(Board)` from the SQLite recovery query — the sweep only mutates scalar fields (Status/ErrorMessage/RetryCount/UpdatedAt) and never reads the navigations.

File: backend/src/Taskdeck.Infrastructure/Repositories/LlmQueueRepository.cs [hunk 3]
```diff
@@ -309,6 +309,47 @@ private async Task<int> CountByStatusAndCaptureKindAsync(
...
+        // Take into a single bounded query with guaranteed ordering.
+        return await _context.LlmRequests
+            .Include(lr => lr.User)
+            .Include(lr => lr.Board)
+            .Where(lr => lr.Status == RequestStatus.Processing
+                && !EF.Functions.Like(lr.RequestType, CaptureRequestTypeLike)
+                && lr.UpdatedAt <= staleBefore)
+            .OrderBy(lr => lr.UpdatedAt)
+            .Take(limit)
+            .ToListAsync(cancellationToken);
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Infrastructure/Repositories/LlmQueueRepository.cs:350]: Fixed in `6cab6881`: same `.Include` removal applied to the non-SQLite (EF) path.

File: backend/src/Taskdeck.Api/Workers/LlmQueueToProposalWorker.cs [hunk 4]
```diff
@@ -138,6 +149,72 @@ private async Task ProcessBatchAsync(CancellationToken ct)
...
+        {
+            // Recovery counts against the retry budget so a request that repeatedly crashes the worker
+            // eventually fails permanently instead of looping forever. MarkAsFailed (Processing -> Failed,
+            // RetryCount++) then ResetForRetry (Failed -> Pending) reuses the same transitions and budget
+            // check as HandleFailureWithRetryAsync, keeping recovery and in-band failure consistent.
+            var budgetRemains = item.RetryCount + 1 < _settings.MaxRetries;
+            item.MarkAsFailed("Recovered from a stale Processing state; the worker that claimed this request did not complete it.");
+            if (budgetRemains)
+            {
+                item.ResetForRetry();
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Api/Workers/LlmQueueToProposalWorker.cs:220]: Good catch — fixed in `6cab6881`. `ProcessSingleItemAsync` now calls `GetBySourceReferenceAsync(ProposalSourceType.Queue, item.Id)` before invoking the planner and, if a Queue-sourced proposal already exists, marks the request `Completed` without reprocessing (mirrors the capture path's existing-proposal short-circuit). This covers both the recovery-requeue path and the pre-existing planner-success-then-save-fail retry. Verified by `ProcessSingleItem_WithExistingQueueProposal_CompletesWithoutReprocessing` (planner set to fail; asserts status Completed + planner CallCount==0).

**Comment by Chris0Jeky** [comment]
## Bot review round — all findings fixed (`6cab6881` + `a31bb83a`)

Addressed every Gemini + Codex finding; all four inline threads resolved with fix replies.

| Bot | Sev | Finding | Resolution |
|-----|-----|---------|------------|
| Gemini | HIGH | Recovery sweep could abort the whole tick on a DB exception. | **Fixed** `efa0e2d0` — sweep wrapped in its own try/catch (re-throws `OperationCanceledException` on shutdown; logs + continues otherwise). |
| Gemini | MED | Unnecessary `.Include(User)/.Include(Board)` in the SQLite recovery query. | **Fixed** `6cab6881` — removed; sweep only mutates scalar fields. |
| Gemini | MED | Same `.Include` on the non-SQLite path. | **Fixed** `6cab6881` — removed. |
| Codex | P2 | **Duplicate proposals**: a worker that created+committed the proposal then crashed before `MarkAsCompleted` would, on recovery-requeue, re-run the planner → a second `PendingReview` proposal (creation isn't idempotent on `SourceReferenceId`). | **Fixed** `6cab6881` — `ProcessSingleItemAsync` now checks `GetBySourceReferenceAsync(Queue, item.Id)` before planning and completes the request if a proposal already exists (mirrors the capture path's guard). Also closes the **pre-existing** planner-success-then-save-fail retry duplicate. Test: `ProcessSingleItem_WithExistingQueueProposal_CompletesWithoutReprocessing` (planner forced to fail → asserts `Completed` + planner `CallCount == 0`). |

### Re-verification (local, Release)
- Backend build clean.
- `LlmQueueToProposalWorkerTests` + `LlmQueueRepositoryIntegrationTests` + both `WorkerResilience` suites: **82/82** (+1 idempotency test).
- `ProposalHousekeepingWorker` suite: 10/10 (the shared `FakeAutomationProposalRepository.GetBySourceReferenceAsync` change is safe).

The duplicate-proposal fix is the notable one: it brings the non-capture drain to parity with the capture path's existing idempotency guard, so neither recovery nor in-band retry can produce duplicate queue proposals.

File: backend/src/Taskdeck.Api/Workers/LlmQueueToProposalWorker.cs [hunk 5]
```diff
@@ -138,6 +164,80 @@ private async Task ProcessBatchAsync(CancellationToken ct)
...
+            // eventually fails permanently instead of looping forever. MarkAsFailed (Processing -> Failed,
+            // RetryCount++) then ResetForRetry (Failed -> Pending) reuses the same transitions and the same
+            // pre-increment budget check (RetryCount + 1 < MaxRetries) as HandleFailureWithRetryAsync. It
+            // deliberately omits that path's IsTransientFailure gate: a stale-Processing row means the
+            // worker crashed mid-flight (no error code to classify), which we always treat as retryable.
+            // The lease itself spaces successive recoveries of a poison-pill payload >= the lease apart, so
+            // no explicit retry backoff is needed here.
+            var budgetRemains = item.RetryCount + 1 < _settings.MaxRetries;
+            item.MarkAsFailed("Recovered from a stale Processing state; the worker that claimed this request did not complete it.");
+            if (budgetRemains)
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Api/Workers/LlmQueueToProposalWorker.cs:218]: Sharp follow-up — fixed in `8cba0723`. `RecoverStuckProcessingItemsAsync` now checks `GetBySourceReferenceAsync(ProposalSourceType.Queue, item.Id)` at the top of the loop, *before* the requeue/fail decision: if a proposal already exists the request actually succeeded (only the completing status flip was lost), so it is `MarkAsCompleted` regardless of retry budget — so a last-attempt row whose proposal was already committed is completed, not failed. New `recovered_stuck_completed` telemetry outcome. Covered by `RecoverStuck_ProposalAlreadyExists_CompletesRegardlessOfBudget` (Theory over budget-remaining and budget-exhausted).

**Comment by Chris0Jeky** [comment]
## Bot round 2 — fixed (`8cba0723`)

**Codex P2 (follow-up):** recovery on a stuck request's *last* attempt marked it `Failed` even when the crashed worker had already committed its proposal — bypassing the drain's existing-proposal guard and mislabeling a successful request.

**Fix:** `RecoverStuckProcessingItemsAsync` now checks `GetBySourceReferenceAsync(Queue, item.Id)` at the top of the loop and `MarkAsCompleted`s the request **regardless of retry budget** when a proposal already exists, ahead of the requeue/fail decision (new `recovered_stuck_completed` telemetry outcome). Thread resolved.

**Verification (local, Release):** build clean; worker + repo + resilience suites **84/84** (+2 from `RecoverStuck_ProposalAlreadyExists_CompletesRegardlessOfBudget`, covering budget-remaining and budget-exhausted). CI green on the prior push; re-running on `8cba0723`.
